File: 29Items-Projects/7Algorithms/5-DIVIDE-CONQUER/src/main/python/divide_conquer/algorithms/karatsuba.py
"""Karatsuba multiplication."""


def multiply(x: int, y: int) -> int:
    """Multiply two integers with Karatsuba's divide-and-conquer strategy."""
    if x < 0 or y < 0:
        sign = -1 if (x < 0) ^ (y < 0) else 1
        return sign * multiply(abs(x), abs(y))
    if x < 10 or y < 10:
        return x * y

    digits = max(len(str(x)), len(str(y)))
    half = digits // 2
    base = 10**half

    high_x, low_x = divmod(x, base)
    high_y, low_y = divmod(y, base)

    z0 = multiply(low_x, low_y)
    z2 = multiply(high_x, high_y)
    z1 = multiply(low_x + high_x, low_y + high_y) - z2 - z0

    return z2 * base * base + z1 * base + z0

File: 29Items-Projects/7Algorithms/5-DIVIDE-CONQUER/src/main/java/com/divideconquer/algorithms/Karatsuba.java
package com.divideconquer.algorithms;

public final class Karatsuba {
    private Karatsuba() {
    }

    public static long multiply(long x, long y) {
        if (x < 0 || y < 0) {
            boolean negative = (x < 0) ^ (y < 0);
            long result = multiply(Math.abs(x), Math.abs(y));
            return negative ? -result : result;
        }
        if (x < 10 || y < 10) {
            return x * y;
        }

        int digits = Math.max(decimalDigits(x), decimalDigits(y));
        int half = digits / 2;
        long base = pow10(half);

        long highX = x / base;
        long lowX = x % base;
        long highY = y / base;
        long lowY = y % base;

        long z0 = multiply(lowX, lowY);
        long z2 = multiply(highX, highY);
        long z1 = multiply(lowX + highX, lowY + highY) - z2 - z0;

        return z2 * base * base + z1 * base + z0;
    }

    private static int decimalDigits(long value) {
        int digits = 1;
        while (value >= 10) {
            value /= 10;
            digits++;
        }
        return digits;
    }

    private static long pow10(int exponent) {
        long result = 1;
        for (int i = 0; i < exponent; i++) {
            result *= 10;
        }
        return result;
    }
}
